### PR TITLE
[dattri.metrics] Add brittleness test

### PR DESCRIPTION
**Pull request by SeanZh30**
### 1. Motivation and Context

This PR is to add brittleness test in metrics. [Reference paper](https://arxiv.org/pdf/2202.00622). This version does not use interpolate.

### 2. Summary of the change

1. Add brittleness.py under metrics. (After review, we may make it wrapped as a function in metric.py.
2. Add mnist_lr_brittleness.py in experiment folder to show how to use this metric to evaluate Influence Function.

### 3. What tests have been added/updated for the change?
 N/A:

### CODE REVIEW

File: dattri/metrics/britteness.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+    model = train_func(new_train_loader)
+    model.to(device)
+    model.eval()
+
+    x, label = test_data
+    x, label = x.to(device), label.to(device)
+
+    output = model(x)
+    pred = output.argmax(dim=1, keepdim=True)
```
Comment (TheaperDeng) [inline dattri/metrics/britteness.py:113]: we may use the `eval_func` to make this support more situations.

File: dattri/metrics/britteness.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    Args:
+        test_data (tuple): The test data consisting of an image and its label.
+        train_loader (DataLoader): DataLoader for the training data.
+        train_data_indices (list[int]):
+            List of indices indicating orignial data the used training data.
+        indices_to_remove (list[int]):
+            Indices of the training data to remove.
+        train_func (function): Function to train the model on the modified dataset.
+        device (torch.device): The device for computation.
+        batch_size (int): Batch size for the DataLoader of the new training set.
```
Comment (TheaperDeng) [inline dattri/metrics/britteness.py:93]: should we add a retraining_num here to do the retraining multiple times to handle the variance?

**Review by TheaperDeng — COMMENTED** [review]
please find some comments below.

File: dattri/metrics/britteness.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+def brittleness(
+    test_data: Tuple[torch.Tensor, torch.Tensor],
+    train_loader: DataLoader,
+    train_data_indices: torch.Tensor,
+    scores: torch.Tensor,
+    steplength: Tuple[int, int, int],
+    train_func: Callable[[DataLoader], torch.nn.Module],
+    device: torch.device,
+    batch_size: int = 64,
+) -> Optional[int]:
```
[the remark this answers is not in the document]
Comment (SeanZh30) [reply dattri/metrics/britteness.py:26]: Fixed.

File: dattri/metrics/britteness.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        batch_size (int): Batch size for the DataLoader of the new training set.
+
+    Returns:
+        bool: True if the prediction flips, otherwise False.
+    """
+    dataset = train_loader.dataset
+    dataset = Subset(dataset, train_data_indices)
+
+    remaining_indices = list(set(range(len(dataset))) - set(indices_to_remove))
+    subset = Subset(dataset, remaining_indices)
```
[the remark this answers is not in the document]
Comment (SeanZh30) [reply dattri/metrics/britteness.py:102]: Fixed.

File: dattri/metrics/britteness.py (hunk 1 above)
Comment (SeanZh30) [reply dattri/metrics/britteness.py:113]: Fixed.

File: dattri/metrics/britteness.py (hunk 2 above)
Comment (SeanZh30) [reply dattri/metrics/britteness.py:93]: When we call the train_mnist or our train function in dattri, there will be a fixed seed (default as 0). In the experiment file, I tried it multiple times and the result seems not change. But for users cutomize train function, we can discuss whether we can add a parameter for retraining or use interpolate.

Comment (TheaperDeng) [reply dattri/metrics/britteness.py]: You mean https://github.com/TRAIS-Lab/dattri/blob/main/dattri/benchmark/datasets/mnist/mnist_lr.py#L18-L53 the `seed` parameter here does not work? Do you mean the logits keep the same or just the binary pred==label.

File: dattri/metrics/britteness.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    model = train_func(new_train_loader)
+    model.to(device)
+    model.eval()
+
+    for test_data in test_loader:
+        x, label = test_data
+        x, label = x.to(device), label.to(device)
+        output = model(x)
+    pred = eval_func(output)
```
Comment (TheaperDeng) [inline dattri/metrics/britteness.py:112]: I think a redesign to the signature of `eval_func` I think maybe let it can be asked to have the same signature as the train_func?, i.e., takes a test_loader and output the predict?

File: dattri/metrics/britteness.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    Returns:
+        bool: True if prediction flips after data removal, otherwise False.
+    """
+    original_sampler = train_loader.sampler
+    original_indices = original_sampler.indices
+    remaining_indices = list(set(original_indices) - set(indices_to_remove))
+
+    new_train_loader = torch.utils.data.DataLoader(
+        train_loader.dataset,
+        batch_size=64,
```
Comment (TheaperDeng) [inline dattri/metrics/britteness.py:100]: maybe use train_loader.batch_size?

File: experiments/mnist_lr_brittleness.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+        eval_func=eval_func,
+        device=device,
+        search_space=[20,40,60,80,100,120,140,180],
+    )
+    if smallest_k is not None:
+        print(f"The number of removal that can make it flip: {smallest_k}")
+    else:
+        print("No k found that causes a flip.")
+    end_time = time.time()
+    print("Total time for brittleness test: ", end_time - start_time)
```
Comment (TheaperDeng) [inline experiments/mnist_lr_brittleness.py:139]: maybe paste some result here? so that it's easier to see if the implementation give reasonable result.

**Comment by TheaperDeng** [comment]
And you may add brittleness here: https://github.com/TRAIS-Lab/dattri?tab=readme-ov-file#metrics-supported

File: experiments/mnist_lr_brittleness.py (hunk 7 above)
Comment (SeanZh30) [reply experiments/mnist_lr_brittleness.py:139]: image
Here is the result of doing brittleness test on the mnist+mlp as the benchmark. The mlp model is trained on 500 size mnist data. It shows 4 different TDA method: IF, RPS, TRAK, TracIn.

**Comment by SeanZh30** [comment]
> And you may add brittleness here: https://github.com/TRAIS-Lab/dattri?tab=readme-ov-file#metrics-supported

Made the PR in [https://github.com/TRAIS-Lab/dattri/pull/121](url)

File: dattri/metrics/britteness.py (hunk 6 above)
Comment (SeanZh30) [reply dattri/metrics/britteness.py:100]: Fixed

**Comment by SeanZh30** [comment]
> And you may add brittleness here: https://github.com/TRAIS-Lab/dattri?tab=readme-ov-file#metrics-supported

Done in this PR

File: dattri/metrics/britteness.py (hunk 5 above)
Comment (SeanZh30) [reply dattri/metrics/britteness.py]: Fixed

**Review by jiaqima — APPROVED** [review]
LGTM

**Review by TheaperDeng — COMMENTED** [review]
Some after-merging comments just to record some possible future work.
